### PR TITLE
Add the argument `dtype` (default to `float32`) to the noise

### LINKED ISSUE #1278: [Bug]: Action noise changes the action dtype
### 🐛 Bug

It's more an unwanted behaviour than a bug. As gym/torch works with `float32`, action noise should be `float32`.

### To Reproduce

```python
import numpy as np

from stable_baselines3.common.env_util import make_vec_env
from stable_baselines3.common.noise import OrnsteinUhlenbeckActionNoise

noise = OrnsteinUhlenbeckActionNoise(np.array([0]), sigma=np.array([0]))
env = make_vec_env("Pendulum-v1")

print(env.action_space.sample().dtype)
print((env.action_space.sample() + noise()).dtype)
```

### Relevant log output / Error message

```shell
float32
float64
```

### System Info

- OS: macOS-13.0.1-arm64-arm-64bit Darwin Kernel Version 22.1.0: Sun Oct  9 20:14:30 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T8103
- Python: 3.9.16
- Stable-Baselines3: 1.8.0a1
- PyTorch: 1.13.1
- GPU Enabled: False
- Numpy: 1.24.0
- Gym: 0.21.0

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] I have provided a minimal working example to reproduce the bug
- [X] I've used the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

### PR DESCRIPTION
**Pull request by sidney-tio**
## Description
Fixed noise to return float32, instead of the default float64, to be consistent with gym action.

Changed all outputs of noise classes to float32, by casting with `np.astype(np.float32)`. 
## Motivation and Context
Closes #1278 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [ ] I have opened an associated PR on the [SB3-Contrib repository](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib) (if necessary)
- [ ] I have opened an associated PR on the [RL-Zoo3 repository](https://github.com/DLR-RM/rl-baselines3-zoo) (if necessary)
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)
- [x] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

File: stable_baselines3/common/noise.py
Comment (araffin) [inline stable_baselines3/common/noise.py:84]: I would add the dtype in the constructor, defaulting to `np.float32` (so it can also work with other dtypes)

File: tests/test_deterministic.py
Comment (araffin) [inline tests/test_deterministic.py:20]: we need to check in the doc that we always use a numpy array as input, otherwise, it might break some codes

Comment (sidney-tio) [reply tests/test_deterministic.py:20]: I did a check on the docs and found numpy arrays used in instances where `NormalActionNoise` is called. For now, it should be fine. 

In the type hints, it's also `np.ndarray` expected for both parameters. Since `np.random.normal` accepts both floats and numpy arrays, the current `__call__` will run regardless.
https://github.com/DLR-RM/stable-baselines3/blob/bea3c44ba52278ec755af0179859b04ab80cdcaf/stable_baselines3/common/noise.py#L27-L41

With the introduction of `.astype`, code that uses floats will fail; perhaps another PR to cast python floats to numpy arrays?

Comment (araffin) [reply tests/test_deterministic.py:20]: > With the introduction of .astype, code that uses floats will fail; perhaps another PR to cast python floats to numpy arrays?

should be fine if we don't have example in the doc, and the type annotations were corrects.
It might break some code but should be very limited (1D action space problems).

**Comment by qgallouedec** [comment]
I'll review tomorrow

File: stable_baselines3/common/noise.py
Comment (qgallouedec) [inline stable_baselines3/common/noise.py:116]: I would drop dtype in `VectorizedActionNoise`, since the type is already specified in `base_noise`. It could lead to unintuitive thing, such as:

```python
from stable_baselines3.common.noise import VectorizedActionNoise, NormalActionNoise
import numpy as np

noise = NormalActionNoise(np.array([0.0, 0.0]), np.array([1.0, 1.0]), dtype=np.float64)
vec_noise = VectorizedActionNoise(noise, 4)
vec_noise().dtype # np.float32
```

Comment (sidney-tio) [reply stable_baselines3/common/noise.py]: Thanks for the feedback!

File: stable_baselines3/common/noise.py
Comment (araffin) [inline stable_baselines3/common/noise.py:35]: minor: missing dtype in docstring

**Review by qgallouedec — APPROVED** [review]

**Comment by qgallouedec** [comment]
lgtm, @araffin if you want to take a look before I merge.

File: docs/misc/changelog.rst
Comment (araffin) [inline docs/misc/changelog.rst:29]:
```suggestion
- Added the argument ``dtype`` (default to ``float32``) to the noise for consistency with gym action (@sidney-tio)
```

**Review by araffin — APPROVED** [review]
LGTM too =)

**Comment by qgallouedec** [comment]
Thanks for contributing @sidney-tio!

**Pull request merged**
